src/Database/LSMTree/Internal/Chunk.hs

```diff
@@ -0,0 +1,98 @@
+module Database.LSMTree.Internal.Chunk
+(
+    -- * Chunks
+    Chunk (Chunk),
+
+    -- * Balers
+    Baler,
+    createBaler,
+    feedBaler,
+    readBalerRemnant
+)
+where
+
+import           Prelude hiding (concat, length, splitAt)
+
+import           Control.Monad.ST.Strict (ST)
+import           Data.STRef.Strict (STRef, newSTRef, readSTRef, writeSTRef)
+import           Data.Vector.Primitive (Vector, concat, length, slice, splitAt)
+import           Data.Word (Word8)
+
+-- * Chunks
+
+-- | A chunk of bytes, typically output during incremental index serialisation.
+newtype Chunk = Chunk (Vector Word8)
+
+-- * Balers
+
+-- Mutable state of a baler.
+data BalerState = BalerState
+                      ![Vector Word8] -- Reverse list of queued blocks
+                      !Int            -- Total size of queued blocks
+
+{-|
+    An object that receives blocks of bytes and repackages them into chunks of
+    preferably a given maximum size.
+-}
+data Baler s = Baler
+                   !Int                  -- Maximum chunk size
+                   !(STRef s BalerState) -- Reference to the mutable state
+
+-- | Creates a new baler.
+createBaler :: Int            -- ^ Maximum chunk size
+            -> ST s (Baler s) -- ^ Creation of the baler
+createBaler maxChunkSize = Baler maxChunkSize <$> newSTRef (BalerState [] 0)
+
+{-|
+    Feeds a baler a block of bytes.
+
+    Bytes received by a bailer are generally queued for later output, but if
+    feeding new bytes makes the queued content exceed the maximum chunk size
+    then as many chunks of maximum size as possible are output.
+-}
+feedBaler :: Vector Word8 -> Baler s -> ST s [Chunk]
+feedBaler block (Baler maxChunkSize stateRef) = do
+    BalerState remnantBlocksRev remnantSize <- readSTRef stateRef
+    let
+
+        totalSize :: Int
+        totalSize = remnantSize + length block
```

Comment:
Does GHC really not figure out that `totalSize` will definitely be used? After all, `totalSize` is compared to `maxChunkSize` in the condition of the `if` expression that immediately follows. I’m not at all against ensuring that variables like `totalSize` get evaluated promptly, but I’d like to avoid spreading `!`-annotations beyond necessity to ensure that the code stays as clear as possible.

Reply:
GHC *might* figure that out, but in the case of most GHC optimisations it is a good rule of thumb is to not trust that GHC will *always* fire optimisations, even if it looks like a simple one. It is better to be explicit

Reply:
Changed in 9c8dec880f539cc47027a16e5872696bc4240850. 🙂